pkg/server/autoconfig/task_markers.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package autoconfig
+
+import (
+	"bytes"
+	"context"
+	"strconv"
+
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/util/encoding"
+	"github.com/cockroachdb/errors"
+)
+
+// InfoKeyCompletionPrefix is the prefix of the key inserted in job_info
+// when a task has completed.
+var InfoKeyCompletionPrefix = roachpb.Key("completed-")
+
+// InfoKeyStartPrefix is the prefix of the key inserted in job_info
+// when a task has started.
+var InfoKeyStartPrefix = roachpb.Key("started-")
+
+// DecodeTaskID returns the task ID for the given infoKey.
+func DecodeTaskID(prefix roachpb.Key, infoKey []byte) (uint64, error) {
+	if !bytes.HasPrefix(infoKey, []byte(prefix)) {
+		return 0, errors.AssertionFailedf("programming error: prefix %q missing: %q", string(prefix), string(infoKey))
+	}
+	infoKey = infoKey[len(prefix):]
+	_, v, err := encoding.DecodeUvarintAscending(infoKey)
+	return v, err
+}
+
+// MakeTaskInfoKey creates a job_info infoKey for the given task.
+func MakeTaskInfoKey(prefix roachpb.Key, taskID uint64) roachpb.Key {
+	infoKey := make([]byte, 0, len(prefix)+10)
+	infoKey = append(infoKey, []byte(prefix)...)
+	infoKey = encoding.EncodeUvarintAscending(infoKey, taskID)
+	return infoKey
+}
+
+// writeStartMarker writes a start marker for the given task ID and
+// also writes its job ID into the value part.
+func writeStartMarker(
+	ctx context.Context, txn isql.Txn, execCfg *sql.ExecutorConfig, jobID jobspb.JobID, taskID uint64,
+) error {
+	startMarkerInfoKey := MakeTaskInfoKey(InfoKeyStartPrefix, taskID)
+	// We cannot use the InfoStorage directly here because this function
+	// is called from two different jobs (the runner and the task) but
+	// must always write with the job ID of the runner to job_info.
```

Comment:
@knz maybe I'm missing something but can we not initialize an `InfoStorage` after loading the job with `jobs.AutoConfigRunnerJobID`. So something like:

```
job, err := jobRegistry.LoadJobWithTxn(ctx, jobs.AutoConfigRunnerJobID, txn)
if err != nil {
    return err
}
infoStorage := job.InfoStorage(txn)
infoStorage.Write(startMarkerInfoKey, []byte(strconv.FormatUint(uint64(jobID), 10)))
```

Our `Write` method on `InfoStorage` under the hood, is a delete of the existing row corresponding to the `job_id, info_key` and an insert of the new row. The reasons for this are explained here - https://github.com/cockroachdb/cockroach/pull/82638/files#diff-d65357c15cffdb009c1b3fd0f6456265d25b94758f326207c531b8f734b59164R328-R348. Different parts of the job system are going to rely on their being exactly one row corresponding to a `<job_id, info_key>` so I'd like to discourage direct SQL manipulation of the `job_info` table.

Reply:
> can we not initialize an InfoStorage after loading the job

I would like to avoid re-loading the permanent job over and over in every task. It's also silly given that the job entry (and its payload, and its progress) are not needed here, just its ID to populate the job_info table.

Reply:
makes sense, I just realized that you might not need to load it. You can just:

```
j := &jobs.Job{id: jobs.AutoConfigRunnerJobID}
infoStorage := j.InfoStorage(txn)
...
```